contracts/Controller.sol

```diff
@@ -91,8 +97,10 @@ contract Controller is Ownable {
      * @dev can only be called when system is not paused
      * @param _actions array of actions arguments
      */
-    // function operate(Actions.ActionArgs[] memory _actions) external isNotPaused {
-    // }
+    function operate(Actions.ActionArgs[] memory _actions) external isNotPaused nonReentrant {
+        _runActions(_actions);
+        _verifyFinalState(_actions[0].owner, _actions[0].vaultId);
```

Comment:
The 0th action may now always have an owner and vaultId, example you could call exercise first and then call a bunch of other actions. You need to get the first action that has a vault owner and id from runActions and pass that into verifyFinalState

Reply:
Great catch, so we definitely need to return that from the `_verifyFinalState`. // cc @antoncoding

Reply:
updated this.